raft/src/main/java/org/apache/kafka/raft/internals/BatchAccumulator.java

```diff
@@ -194,14 +196,45 @@ private void completeCurrentBatch() {
         MemoryRecords data = currentBatch.build();
         completed.add(new CompletedBatch<>(
             currentBatch.baseOffset(),
-            currentBatch.records(),
+            Optional.of(currentBatch.records()),
             data,
             memoryPool,
             currentBatch.initialBuffer()
         ));
         currentBatch = null;
     }
 
+    public void appendLeaderChangeMessage(LeaderChangeMessage leaderChangeMessage, long currentTimeMs) {
+        maybeCompleteDrain();
+        ByteBuffer buffer = memoryPool.tryAllocate(256);
+        if (buffer != null) {
+            MemoryRecords data = MemoryRecords.withLeaderChangeMessage(
+                    this.nextOffset, 
+                    currentTimeMs, 
+                    this.epoch, 
+                    buffer, 
+                    leaderChangeMessage);
+            completed.add(new CompletedBatch<>(
+                nextOffset,
+                Optional.empty(),
+                data,
+                memoryPool,
+                buffer
+            ));
+            nextOffset += 1;
+        }
+    }
+
+    public void flush() {
```

Comment:
I don't think we should expose this functionality. I think users of these type will always call `flush` after calling `appendLeaderChangeMessage`. If so why not do that implicitly in the method.

Reply:
https://github.com/apache/kafka/pull/10480#discussion_r610237424

Reply:
@jsancio